src/declarations/stencil-private.ts

```diff
@@ -850,12 +850,32 @@ export interface ComponentCompilerPropertyComplexType {
   references: ComponentCompilerTypeReferences;
 }
 
+/**
+ * A record of `ComponentCompilerTypeReference` entities.
+ *
+ * Each key in this record is intended to be the names of the types used by a component. However, this is not enforced
+ * by the type system (I.E. any string can be used as a key).
+ *
+ * Note any key can be a user defined type or a TypeScript standard type.
+ */
 export interface ComponentCompilerTypeReferences {
-  [key: string]: ComponentCompilerTypeReference;
+  [typeName: string]: ComponentCompilerTypeReference;
 }
```

Comment:
I think it can go either way - with a `Record` type, it's cleaner looking (IMO), but comes at the cost of hiding some of the underlying detail. With `Record<string, ComponentCompilerTypeReference>` (or any types really) we don't have the opportunity to name the key in the index signature.  We _could_ argue that we 'lose' some data that way for consumers.  However, we now have JSDoc on this type, so whether or not we're _really_ losing anything probably comes down to how various editors/plugins show the data from the language server.

FWIW this is what JetBrains shows me (there's no index signature at all)
![Screen Shot 2022-04-18 at 12 03 24 PM](https://user-images.githubusercontent.com/1930213/163836637-456c790a-ba8a-48ad-827d-1951ca8d2b66.png).

The only other thing I can think of is that moving this to a `Record` would require to change this from an `interface` to a `type`:
```
export type ComponentCompilerTypeReferences = Record<string, ComponentCompilerTypeReference>;
```

AFAICT, we don't export this type anywhere (either directly or indirectly), and it's defined in `stencil-private.ts` (so nobody should be using it other than us). That minimizes the chances of anyone relying on the existing `ComponentCompilerTypeReferences` def as an interface - by moving this from an `interface` to a `type`, we remove the possibility of type merging/augmenting `ComponentCompilerTypeReference` automagically.

With that in mind, I think I'd be happy making the change. Worst case scenario, its an easy enough revert. 

93a06af3b